### PR TITLE
Fix int32_t microseconds overflow in navigation.

### PR DESCRIPTION
**Pull request by pieniacy**
## Bug and reproduction
There is a severe bug in navigation. To reproduce: after 40mins of waiting arm the quad in the althold mode and the **motors will instantly spin at full throttle**. Needless to say, do not try that with the props on (and don't ask how I found out about this issue).

## Cause
The navigation uses something called `deltaMicrosPositionUpdate` and it is calculated from zero at first update or since last update. This delta can get big either if you wait before first arm or between arms. This value is held in `int32_t` type, and 2^31 microseconds equals ~36 minutes, so in the range from ~36 to ~72 minutes **the value will overflow** to a negative number. There is a check if the delta is not too big, but a negative number obviously always satisfies this check. Then, the negative value of delta is taken into computations causing a mess and causing altitude controller to output full throttle.

To be precise `deltaMicrosPositionUpdate` is of type `timeDelta_t` and its definition [looks like so](https://github.com/iNavFlight/inav/blob/c26a676bb1df029cb859b51875898c58d7279134/src/main/common/time.h#L27):
```
// time difference, 32 bits always sufficient
typedef int32_t timeDelta_t;
```
Well, not always sufficient.

## Naive solution
The naive solution would be to just change `timeDelta_t` to `int64_t`, but as this type is widely used in the project this would introduce performance degradation, for example [this](https://godbolt.org/z/3vrna15Pn) shows that 32 bits value of `timeDelta_t` is handled by the FPU while 64 bits is delegated to the soft emulation!
Almost all uses of `timeDelta_t` deal with very small changes which do not present much risk of overflow. Moreover, for example in the scheduler performance is quite important. Therefore just changing `timeDelta_t` to `int64_t` does not seem right.

## This solution
_Important changes are at the top of `time.h` and in lines where `timeDeltaLarge_t` is added in `navigation_*`._

Firstly, the comment of `timeDelta_t` was changed to accurately reflect the capabilities of the type and a `TIMEDELTA_MAX` define was added. Secondly there is a new `timeDeltaLarge_t` which typedefs to `int64_t`, which overflows at 300 000 years, so this one can definitely can hold any practical interval.

Secondly, in `navigation_fixedwing.c`, `navigation_multicopter.c` and `navigation_rover_boat.c` appropriate `deltaMicrosPositionUpdate`s are changed to `timeDeltaLarge_t` and since they are checked for maximum possible interval anyway they will not be passed as `timeDelta_t` if they are big enough to cause overflow.

The last somewhat important change is in `navigation_private.h` where the maximum interval is asserted to fit in the `timeDelta_t`.

## By the way
I also looked at other occurences of `timeDelta_t` to figure out where overflow is possible/dangerous and apart from navigation I haven't found any cases requiring more than 32 bits, so I only did some minor refactors of related code here and there.

### CODE REVIEW

**Comment by pieniacy** [comment]
It seems CI build failed to setup its environement (compiler download failed) and write access to the repository is required to rerun checks, can I have a rerun please?

**Comment by JulioCesarMatias** [comment]
very good, but replacing `int32_t` with `uint32_t` would be enough, wouldn't it?

**Comment by JulioCesarMatias** [comment]
good job, but replacing `in32_t` with `uint32_t` would be better, wouldn't it?

File: src/main/navigation/navigation_rover_boat.c [hunk 1]
```diff
@@ -71,11 +71,11 @@ void applyRoverBoatPositionController(timeUs_t currentTimeUs)
     static timeUs_t previousTimePositionUpdate;         // Occurs @ GPS update rate
     static timeUs_t previousTimeUpdate;                 // Occurs @ looptime rate
 
-    const timeDelta_t deltaMicros = currentTimeUs - previousTimeUpdate;
+    const timeDeltaLarge_t deltaMicros = currentTimeUs - previousTimeUpdate;
```
Comment (digitalentity) [inline src/main/navigation/navigation_rover_boat.c:74]: I don't see how this fixes the problem. We can still get negative `deltaMicros`, just much later. I would like to avoid 64-bit math as it's expensive in terms of CPU cycles. A better mitigation would be to ensure `deltaMicros` is positive.

**Review by digitalentity — CHANGES_REQUESTED** [review]

File: src/main/navigation/navigation_rover_boat.c (hunk 1 above)
Comment (pieniacy) [reply src/main/navigation/navigation_rover_boat.c:74]: No worries, I will explain my thought process a little bit better.

---

> We can still get negative `deltaMicros`

I disagree. Assuming signed type negative `deltaMicros` can be a result of:
1. `currentTimeUs < previousTimeUpdate` - this case will not happen due to semantics of `previousTimeUpdate` with a simple assumption that `currentTimeUs` is non-decreasing
2. `TYPE_MAX < (currentTimeUs - previousTimeUpdate) < (2 * TYPE_MAX)` (for simplicity we are considering only the first interval) - this is the case that can happen and happened to me. With 32-bit type the problem pops up in the interval [~36&nbsp;mins,&nbsp;~72 mins], which is an interval that can easily be seen in a perfectly normal use case. On the other hand 64-bit type gives us interval of failure [~300000&nbsp;years,&nbsp;~600000 years] and I don't believe it is sane to talk about the craft being on for 300000&nbsp;years. (fun fact: 300000&nbsp;years ago Homo Sapiens started to apppear on Earth)

Therefore I believe we cannot get negative `deltaMicros`.

---

> avoid 64-bit math as it's expensive in terms of CPU cycles

Note that any function taking deltaMicros into any serious calculation (like `update2DPositionHeadingController` or `updateAltitudeVelocityController_MC`) still uses `timeDelta_t`, so performance there is not changed at all. Changes from 32 to 64-bit variables are very selective and after making sure it fits the delta is converted to 32-bit (implicitly, because of function's parameter type). And the check if it fits does not even have to be added. It is already being done, because delta is always checked if it is not too big and not too big value is asserted to fit in the 32-bit type.
To summarize: only a few more cpu cycles will be used during the first comparison, but this amount is small and necessary. All&nbsp;serious calculations are left unchanged.

---

> A better mitigation would be to ensure `deltaMicros` is positive.

I disagree. After ~72 mins the delta held in 32-bit type would indicate a positive value again, but this value will be wrong (because it wraped around) and can falsely show for example 1us when it should show 4294967297us. This can lead to wrongly satisfying all checks and proceeding with calculations without resetting appropriate controllers after a ~72min delay. So&nbsp;I&nbsp;believe this one would be a solution that does not actually solve the problem, just pushes it by half an hour.

Comment (digitalentity) [reply src/main/navigation/navigation_rover_boat.c:74]: Fair enough, you got me convinced :smile:

**Review by digitalentity — APPROVED** [review]
LGTM
